### PR TITLE
Add Lunell

### PR DESCRIPTION
**Pull request by WarriorGallade**

### CODE REVIEW

File: data/mods/gen9ssb/scripts.ts [hunk 1]
```diff
@@ -1085,6 +1085,7 @@ export const Scripts: ModdedBattleScriptsData = {
 					const isCharging = move.flags['charge'] && !this.volatiles['twoturnmove'] &&
 						!(move.id.startsWith('solarb') && ['sunnyday', 'desolateland'].includes(this.effectiveWeather())) &&
 						!(move.id === 'fruitfullongbow' && ['sunnyday', 'desolateland'].includes(this.effectiveWeather())) &&
+						!(move.id === 'praisethemoon' && ['gravity'].includes(this.effectiveWeather())) &&
```
Comment (HoeenCoder) [inline data/mods/gen9ssb/scripts.ts:1088]:
```suggestion
						!(move.id === 'praisethemoon' && this.battle.field.getPseudoWeather('gravity')) &&
```
Gravity is a Pseudo weather, not an actual weather condition.

File: data/mods/gen9ssb/conditions.ts [hunk 2]
```diff
@@ -1227,6 +1239,90 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
 			this.add('-end', pokemon, 'Attract', '[silent]');
 		},
 	},
+
+	gravity: {
```
Comment (HoeenCoder) [inline data/mods/gen9ssb/conditions.ts:1243]: Which parts of gravity are modified? I see several methods that appear to be copy/pasted from base code. You are inheriting from the existing gravity condition already, so theres no need to include a method here unless you are overriding it to change how it works.

File: data/mods/gen9ssb/conditions.ts [hunk 3]
```diff
@@ -1227,6 +1239,90 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
...
+		onFieldResidualSubOrder: 2,		
+		onFieldEnd() {
+			this.add('-fieldend', 'move: Gravity');
+			const leftstate = this.p1.active[0];
+			const rightstate = this.p2.active[0];
+			const lunellonleft = (leftstate && leftstate.isActive && leftstate.name == "Lunell");
+			const lunellonright = (rightstate && rightstate.isActive && rightstate.name == "Lunell");
+			if (lunellonleft || lunellonright) {
+				this.add(`c:|${getName('Lunell')}|ope there goes gravity`);
+			}
```
Comment (HoeenCoder) [inline data/mods/gen9ssb/conditions.ts:1322]:
```suggestion
			const activePokemon = this.getAllActive();
			for (let a of activePokemon) {
				if (a.name === "Lunell") {
					this.add(`c:|${getName('Lunell')}|ope there goes gravity`);
					break;
				}
			}
```
This is a cleaner way to implement (unless I'm wrong about `this` being an object of `Battle`).

File: data/mods/gen9ssb/abilities.ts [hunk 4]
```diff
@@ -699,6 +699,34 @@ export const Abilities: {[k: string]: ModdedAbilityData} = {
 		flags: {},
 	},
 
+	// Lunell
+	lowtidehightide: {
+		shortDesc: "Switch-in sets Gravity, water immune, traps water type foes.",
+		name: "Low Tide, High Tide",
+		onStart(source) {
+			this.field.setWeather('gravity');
```
Comment (HoeenCoder) [inline data/mods/gen9ssb/abilities.ts:707]:
```suggestion
			this.field.addPseudoWeather('gravity', source);
```

File: data/mods/gen9ssb/abilities.ts [hunk 5]
```diff
@@ -699,6 +699,34 @@ export const Abilities: {[k: string]: ModdedAbilityData} = {
 		flags: {},
 	},
 
+	// Lunell
+	lowtidehightide: {
+		shortDesc: "Switch-in sets Gravity, water immune, traps water type foes.",
+		name: "Low Tide, High Tide",
```
Comment (HoeenCoder) [inline data/mods/gen9ssb/abilities.ts:705]: I'm unsure if commas are allowed in ability/item/move names, its a special character.

File: data/mods/gen9ssb/abilities.ts [hunk 6]
```diff
@@ -699,6 +699,34 @@ export const Abilities: {[k: string]: ModdedAbilityData} = {
...
+		onStart(source) {
+			this.field.setWeather('gravity');
+		},
```
Comment (KrisXV) [inline data/mods/gen9ssb/abilities.ts:708]:
```suggestion
		},
```

File: data/mods/gen9ssb/conditions.ts [hunk 7]
```diff
@@ -587,6 +587,18 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
 		innateName: "Regenerator",
 		shortDesc: "User will heal 33% of their max HP on switch-out.",
 	},
+	lunell: {
+		noCopy: true,
```
Comment (KrisXV) [inline data/mods/gen9ssb/conditions.ts:591]:
```suggestion
		noCopy: true,
```

File: data/mods/gen9ssb/conditions.ts [hunk 8]
```diff
@@ -587,6 +587,18 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
...
+		noCopy: true,		
+		onStart() {
+			this.add(`c:|${getName('Lunell')}|vapowo`);
+		},
+		onSwitchOut() {
+			this.add(`c:|${getName('Lunell')}|brb looking for bean images, don't disturb`);
+		},
+		onFaint() {
+			this.add(`c:|${getName('Lunell')}|*sad vaporeon noises*`);
+		},
```
Comment (KrisXV) [inline data/mods/gen9ssb/conditions.ts:600]:
```suggestion
		},
```

File: data/mods/gen9ssb/conditions.ts [hunk 9]
```diff
@@ -1227,6 +1239,90 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
...
+			}
+		},
+		onModifyMove(move, pokemon, target) {
+			if (move.flags['gravity'] && !move.isZ) {
+				this.add('cant', pokemon, 'move: Gravity', move);
+				return false;
+			}
+		},
+		onFieldResidualOrder: 27,
+		onFieldResidualSubOrder: 2,
```
Comment (KrisXV) [inline data/mods/gen9ssb/conditions.ts:1313]:
```suggestion
		onFieldResidualSubOrder: 2,
```

File: data/mods/gen9ssb/conditions.ts [hunk 10]
```diff
@@ -1227,6 +1239,90 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
...
+			}
+		},
+		onFieldResidualOrder: 27,
+		onFieldResidualSubOrder: 2,		
+		onFieldEnd() {
+			this.add('-fieldend', 'move: Gravity');
+			const leftstate = this.p1.active[0];
+			const rightstate = this.p2.active[0];
+			const lunellonleft = (leftstate && leftstate.isActive && leftstate.name == "Lunell");
+			const lunellonright = (rightstate && rightstate.isActive && rightstate.name == "Lunell");
```
Comment (KrisXV) [inline data/mods/gen9ssb/conditions.ts:1319]:
```suggestion
			const lunellonleft = (leftstate?.isActive && leftstate.name === "Lunell");
			const lunellonright = (rightstate?.isActive && rightstate.name === "Lunell");
```

File: data/mods/gen9ssb/conditions.ts (hunk 3 above)
Comment (KrisXV) [inline data/mods/gen9ssb/conditions.ts:1322]:
```suggestion
			}
```

File: data/mods/gen9ssb/conditions.ts [hunk 11]
```diff
@@ -1227,6 +1239,90 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
...
+			if (lunellonleft || lunellonright) {
+				this.add(`c:|${getName('Lunell')}|ope there goes gravity`);
+			}			
+		},	
+	},
+
```
Comment (KrisXV) [inline data/mods/gen9ssb/conditions.ts:1325]:
```suggestion
		},
	},
```

File: data/mods/gen9ssb/moves.ts [hunk 12]
```diff
@@ -1346,6 +1346,46 @@ export const Moves: {[k: string]: ModdedMoveData} = {
 		type: "Psychic",
 	},
 
+	// Lunell
+
```
Comment (KrisXV) [inline data/mods/gen9ssb/moves.ts:1350]:
```suggestion
	// Lunell
```

**Review by HoeenCoder — CHANGES_REQUESTED** [review]
This is a first pass review, I'll take a closer look after this is addressed, thank you for your help, and don't worry too much about the issues, it happens to all of us.

Check and fix the actions failures, you can avoid these by running `npm test` locally.

Also, please make sure to test your code if you are not already (boot the server, try the moves, make sure everything works correctly). If you need help with this, feel free to ask.

File: data/mods/gen9ssb/moves.ts [hunk 13]
```diff
@@ -1346,6 +1346,46 @@ export const Moves: {[k: string]: ModdedMoveData} = {
...
+			}
+			if (!this.runEvent('ChargeMove', attacker, defender, move)) {
+				return;
+			}
+			attacker.addVolatile('twoturnmove', defender);
+			return null;
+		},
+		secondary: null,
+		hasSheerForce: true,
+
```
Comment (KrisXV) [inline data/mods/gen9ssb/moves.ts:1379]:
```suggestion
		hasSheerForce: true,
```

File: data/mods/gen9ssb/scripts.ts (hunk 1 above)
Comment (KrisXV) [inline data/mods/gen9ssb/scripts.ts:1088]: the check for gravity is different from weathers, i will try to find it

File: data/mods/gen9ssb/conditions.ts (hunk 2 above)
Comment (WarriorGallade) [reply data/mods/gen9ssb/conditions.ts:1243]: Yep, most of this is indeed the same as vanilla gravity code, but inheriting doesn't seem to work in testing when I delete the rest of it. The onFieldEnd method at the end is the only one being overridden/modified.

File: data/mods/gen9ssb/moves.ts [hunk 14]
```diff
@@ -1346,6 +1346,45 @@ export const Moves: {[k: string]: ModdedMoveData} = {
 		type: "Psychic",
 	},
 
+	// Lunell
+
+	praisethemoon: {
```
Comment (HisuianZoroark) [inline data/mods/gen9ssb/moves.ts:1351]:
```suggestion
	// Lunell
	praisethemoon: {
```

File: data/mods/gen9ssb/conditions.ts (hunk 2 above)
Comment (HisuianZoroark) [reply data/mods/gen9ssb/conditions.ts:1243]: Yeah unfortunately Gravity is a move-based pseudoWeather so you have to kinda reimplement gravity to make any slight tweaks like this.

File: data/mods/gen9ssb/conditions.ts [hunk 15]
```diff
@@ -1227,6 +1239,89 @@ export const Conditions: {[k: string]: ModdedConditionData & {innateName?: strin
 			this.add('-end', pokemon, 'Attract', '[silent]');
 		},
 	},
+
+	gravity: {
+		inherit: true,
```
Comment (HisuianZoroark) [inline data/mods/gen9ssb/conditions.ts:1244]:
```suggestion
```
Not needed since it isn't being inherited like that

**Review by HisuianZoroark — APPROVED** [review]

**Pull request merged**